src/charm.py

```diff
@@ -94,8 +166,14 @@ def configure(self, event: ops.HookEvent) -> None:
             }
         )
 
-        # TODO: properly start and stop the service depends on the relations to
-        # keystone and grafana-agent.
+        data = self._get_keystone_data()
+        if not data:
+            logger.info("Keystone credentials are not available, stopping services.")
+            snap_service.stop()
+            return
+
+        logger.info("Keystone credentials are available, starting services.")
+        self._write_cloud_config(data)
         snap_service.start()
```

Comment:
We might need a `snap_service.restart()`, I tested manually with the snap: starting an already started service will not take effect

Reply:
oh, it should have been `changes-requested` because it's kind of a bug